api/jobs/task.js

```diff
@@ -18,14 +17,41 @@ class MonitorJob extends job.Job {
      */
     run(db, done) {
         const self = this;
-        const targetHour = (new moment()).hours();
+        /**
+         * Filter tasks to run base on job run time
+         * @param {object} task object from db
+         * @return {boolean} return true if can run now
+         */
+        function tasksFilter(task) {
+            const lastStart = task.start || 0;
+            const lastEnd = task.end || 0;
+            const now = Date.now();
+            const duration = lastEnd - lastStart;
+            if (duration <= 60 * 1000 && duration >= 0) {
+                return true;
+            }
+            if (duration <= 10 * 60 * 1000 && now - lastEnd >= 2 * 60 * 60 * 1000) {
+                return true;
+            }
+            if (duration <= 30 * 60 * 1000 && now - lastEnd >= 4 * 60 * 60 * 1000) {
+                return true;
+            }
+            if (duration <= 60 * 60 * 1000 && now - lastEnd >= 12 * 60 * 60 * 1000) {
+                return true;
+            }
+            if (duration > 60 * 60 * 1000 && now - lastEnd >= 24 * 60 * 60 * 1000) {
+                return true;
+            }
+            return false;
+        }
+
         common.db.collection("long_tasks").find({
             autoRefresh: true,
-            r_hour: targetHour
         }).toArray(function(err, tasks) {
             log.d('Running Task Monitor Job ....');
             log.d("job info:", self._json, tasks);
-            tasks.forEach((task)=>{
+            const filteredTasks = tasks.filter(tasksFilter);
+            filteredTasks.forEach((task) => {
                 return Promise.coroutine(function *() { // eslint-disable-line require-yield
```

Comment:
I think we also need is running state, or else we might run some long running reports multiple times even before they finished.
And just in case we should reset running flag on master process start, just like we do with cohorts

Reply:
In jobs  data schema, we dont have "in_progress" property, instead, we prvoide  "status":https://github.com/Countly/countly-server/blob/master/api/parts/jobs/job.js#L13-L19

So here I can add status checking.

Reply:
I am not talking about the job, but about tasks. They don't have statuses, jobs have.
So what happens if tasks gets run for more then 1 hour?
Next job that will run in 1 hour, will rerun this task again, even though previous one is not yet finished,

Reply:
here we have status for the task, just recheck this, so you won't run the task, that is already running
https://github.com/Countly/countly-server/blob/master/api/utils/taskmanager.js#L361

Reply:
OK updated